File: 78-subsets/subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> temp;
        Helper(0,nums,ans,temp);
        return ans;
    }
    void Helper(int index,vector<int>& nums,vector<vector<int>>& ans,vector<int>& temp){
        if(index==nums.size())
        {
            ans.push_back(temp);
            return;
        }
        // 2 cases if you consider this number 
        temp.push_back(nums[index]);
        Helper(index+1,nums,ans,temp);
        temp.pop_back();
        Helper(index+1,nums,ans,temp);
    }
};

File: 78-subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        Helper(0,nums,ans,temp);
        return ans;
    }
    public void Helper(int i, int[] nums, List<List<Integer>> ans, List<Integer> temp){
    if(i==nums.length){
        ans.add(new ArrayList<>(temp));
        return;
    }
     temp.add(nums[i]);
     Helper(i+1,nums,ans,temp);
     temp.remove(temp.size()-1);
     Helper(i+1,nums,ans,temp);
    }
}